### DISCUSSION TITLE
Just wishes for new features

### DISCUSSION THREAD

**Discussion by Lichtenshtein** [itagagaki/charu3-SE#8] [Ideas]
It's nice to see that someone else is using this great little program and has even taken the time to improve it. Maybe I misunderstood, but if the development will be continued, then here are my wishes for the new features:

- [X] 1. ActivatIng/inserting an object with a single mouse click. Does not apply to folders. It might be worth adding a special mode for editing objects in the Charu window to avoid problems with accidental paste triggering while moving objects.
- [ ] 2. Keep the focus on the last window when you call Charu. It is increasingly common to see input forms that hide when the focus shifts from them (e.g., browser extensions like Selection Search that work in popups). While selecting an object to insert from Charu the focus is lost and you have to reopen these forms all the time. Charu аnalog - CopyQ can keep the cursor focus on such forms when you select an object and pastes text into the current window after activating the object. I do not consider CopyQ as a replacement for Charu...
- [ ] 3. An option to quickly paste the first 10 objects in the history with hotkeys. For example RCtrl+1, RCrl+2 or Alt+1, etc. In Charu you can assign hotkeys to objects, but in the process of filling the history, the numbering of objects in the history is shifted down in the history, so it makes no sense to assign keys to objects in the history. Therefore, i would like to be able to "fix" hotkeys behind any first 10 objects in the history.
- [ ] 4. Burst paste. 
- a) With "burst paste" in FIFO/FILO mode ALL objects are immediately pasted consecutively. It is desirable to be able to choose the type of separator: each object on a new line/space or comma separated/separated by some kind of separator ; === ; between objects/etc.
- b) Sequential one-time insertion of all objects in a row from those checkboxed in Charu.
- [ ] 5. An option to disable the search by first letters in Charu window.
- [ ] 6. An option to quickly activate and paste the first 10 objects in the history by pressing the numbers 1-0 on keyboard while Charu window is opened.
- [X] 7. Automaticaly "unlock" objects copied from specific applications, such as password managers (KeePass, 1Password, etc). Idea is in one-time paste, to not keep passwords in history. Or option to disable adding objects to history for specific programs at all.
- [ ] 8. Option to "move object to the beginning of history" after its insertion.

Please post ready-made binaries for future versions in the releases section, if it's not difficult for you.

**Comment by itagagaki** [comment]
@Lichtenshtein I didn't notice your post until today because I forgot to set up WATCH. My sincere apologies. I will check carefully later.
Thank you very much!!

**Reply by Lichtenshtein** [reply]
Great job, over all. In my small personal experience, the program has become much faster, than original.

**Reply by itagagaki** [reply]
If it is faster, it could be because the delay time value in the settings is zero or very small, or because I have disabled the code to get the text for the $SEL expansion...

**Comment by itagagaki** [comment]
@Lichtenshtein 

> 6. An option to quickly activate and paste the first 10 objects in the history by pressing the numbers 1-0 on keyboard while Charu window is opened.

What should happen if the item is a folder?

**Reply by itagagaki** [reply]
Oh, you mean the items in the history folder?
You can have more than one history folder, in which case you cannot use only the 1-0 key to determine which item in which folder to retrieve.
What should I do?
And is this different from wish number 3?

**Reply by Lichtenshtein** [reply]
> Oh, you mean the items in the history folder?

Yes, yes. My proposal concerns the History folder. With parameter "Number=yes" for a folder, the fragments in the History folder become numbered. This could be used for inserting a fragment simply by entering its digit.

If you think about it, the ability to quickly insert a fragment can be useful for any folder, not just the History folder...

Maybe enable "Number=yes" by default for all folders?

Then each folder will have its own numbering... Accordingly, Charu must take this into account independently and respond to the input of digits according to the numbering within the currently selected folder...

Open folder in Popup -> check the fragment number -> enter the digit -> insert the fragment. Voilà.

I just noticed that the fragments in the History folder are numbered in the order of their appearance in the history. The very first one is at the bottom. I don't remember if there is an option in Charu to do the opposite, so that the newest fragment is always numbered as 1... Maybe add an option for reverse numbering? Reverse=yes or something like that.

> You can have more than one history folder

I completely forgot about it. I don't know why would i need a second History folder...

Ha. It turns out that all these years i had a second History folder in the Tutorial folder. And it simply duplicates the main History folder. The same clipboard fragments are collected in it as in the main one.

Can different History folders work independently of each other?

> in which case you cannot use only the 1-0 key to determine which item in which folder to retrieve.
What should I do?

If the history is common in all History folders, then can we leave only one History folder? Do you use a second History folder for yourself?

Several independent clipboards with different histories - a request to compare to the new functionality. I don't really need something like that. Although, I have seen similar questions in the topics of other clipboard managers. Such a feature in Charu might attract some new users.

> And is this different from wish number 3?

Yes. The difference is that in "number 3" Charu Popup is not called.

I would call it a "Hot Buffer". The idea is that if you still remember what is contained in the last ~10 fragments in the history, opening a Popup for their insertion may be unnecessary. It is easier and faster to call the desired fragment from memory (and history) with a hotkey.

For now it is possible to assign a hotkey to a fragment in the history, but it is useless. In 99% of cases, calling a fragment from the History folder more than once is impractical. The fragment itself moves down the history with the assigned key until it is erased from the history itself. The hotkey is erased too.

Yes, "Hot Buffer"! In the "Hot Buffer", the first ~10 fragments have fixed hotkeys (Alt+1, Alt+2 or RCtrl+1, RCrl+2, etc), which will not move along with the fragment.

**Reply by itagagaki** [reply]
I don't know why Charu allowed users to have multiple history folders. If you have multiple history folders, the contents will vary depending on when the history folder was created. But of course, if you look at the history from newest to oldest, the contents will be the same up to the middle of the history. And you can manually rearrange the order of the data, or edit or delete any data, or add new data. I don't think it makes sense to do that, but it's possible.

So, some rule must be established to determine which folder is this "Hot Buffer".
If we can even decide on this rule, users may be able to designate not only the history folder but also the regular folder as a "Hot buffer".

Oh, and the fact that it is possible to add items to the history folder means that there may be folders instead of data in the first 10 items of the history folders. We have to consider what happens in that case.

**Reply by itagagaki** [reply]
What about the following specifications?

- The keys `MODIFIER`+`1`-`0` are registered as a hotkey to paste one of the first 10 items in the top folder of the tree (not necessarily the history folder).
- A combination of `LShift`, `RShift`, `LCtrl`, `RCtrl`, `LAlt`, `RAlt` can be specified in the configuration as the `MODIFIER` above.
- Since it is a hotkey, it works the same whether the data tree view is popped up or not.
- If the modifier is set to none (empty), the `1`-`0` key will not work for the "select an item by typing" function in the data tree view. (However, registering a hotkey without the modifier will also interfere with normal work in other Windows applications, so you don't want to do that).
- If the first item in the tree is not a folder, nothing happens.
- For example, if the first folder in the tree has only three items, the `4`-`0` key will do nothing.
- If the item pointed to by the key is not data but a folder, nothing happens.

**Reply by Lichtenshtein** [reply]
> I don't think it makes sense to do that, but it's possible.

By logic, the History Folder should be intended for one-time fragments. But in Charu, freedom is originally conceived as an element.

It is possible to set all one-time items in folder "permanent". Even in History Folder. And then change History Folder into a regular folder in folder options. Similarly, it is possible to do the opposite - make another History Folder from a regular folder.

On the one hand, it's great and should stay that way. But on the other hand, as i understood, it makes the implementation of my wishes much more confusing and difficult.

1. Hot Buffer

> So, some rule must be established to determine which folder is this "Hot Buffer".  
> If we can even decide on this rule, users may be able to designate not only the history folder but also the regular folder as a "Hot buffer".

"Hot buffer" makes sense only for the History Folder. Data items inside any folders, except for the History Folder, are by default permanent items. Assigning hotkeys to permanent items is not a problem.

What should happen to the "Hot Buffer" if the History Folder is turned into a regular folder? Maybe just "if regular_folder=true, then hotbuffer=false".

What should happen if you turn a regular folder into a History Folder? Enable "Hot Buffer" for the history folder manually by specifying hotbuffer=true in the folder options.

How should "Hot Buffer" work if there are multiple History Folders and their contents are the same, but still depend on the time of folder creation? hotbuffer=true should only work for the History Folder for which the hotbuffer=true parameter is specified.

> the fact that it is possible to add items to the history folder means that there may be folders instead of data in the first 10 items of the history folders. We have to consider what happens in that case.

Yes, there is a problem with this.

It doesn't make much sense to create folders and one-time items in the folder with one-time items, unless you plan to later convert the History Folder into a regular folder. If that's the case, then you don't need a "Hot Buffer".

It makes no sense to prohibit the user from manually creating folders in the History Folder. Firstly, Charu is not about restrictions. Secondly, if there are too many fragments in the History Folder, they are distributed among automatically created folders. Folders in a disposable folder are created, so the user should be able to create them too.

Therefore, if the item in the History pointed to by the key is not data but a folder, then a) do nothing b) open and show the folder in a popup. I think it makes sense to stop at a).

In summary, it turns out that it should be something like this:

- "Hot Buffer" works only when the hotbuffer=true parameter is present in the properties of the History Folder.
- If hotbuffer=true, the keys MODIFIER+1-0 are registered as a hotkey to paste one of the first 10 items in the History Folder of the tree (necessarily - the history folder).
- A combination of LShift, RShift, LCtrl, RCtrl, LAlt, RAlt can be specified in the configuration as the MODIFIER above.
- The MODIFIERs are global keys and work whether the data tree view is popped up or not.
- If the History Folder in the tree has only three items, the MODIFIER+4-0 key will do nothing.
- If the item pointed to by the key is not data but a folder, nothing happens.

2. Number key pasting

Here arose 2 different proposals, practically performing the same thing, but in different ways.

a) Select and Paste Item in folder by Typing Its Number

At the moment, if you specify Number=yes in the folder properties, then:

- The item number in the folder is assigned only for new items.
- The numbering goes from the old item to the new one.
- The oldest item will have number 1 and will be located at the bottom of the folder.
- If you set a certain number of one-time items "permanent" in the History Folder, the number of items will exceed the set limit (for example, with count=140) and the numbering will start over. The screenshot shows an example of what will happen if the items in the folder exceed the limit of 140 (with count=140).

  ![test1](https://github.com/itagagaki/charu3-SE/assets/13855295/036e2bf4-dd31-4beb-baf7-712b6104e88e)  

- I cleared the History Folder by selecting "Empty this folder" from the context menu. But as you can see in the screenshot, the numbering starts not from the beginning, but continues. I thought that the issue was due to two duplicate history folders, but after deleting the second folder, nothing changed.

  ![test2](https://github.com/itagagaki/charu3-SE/assets/13855295/6d3f94bd-a860-4940-b21f-d018f37aa773)  

_By the way, as you can see, there's a bug with folder naming inside History Folder. For some reason they all called "????"_

In my opinion, the current format of items numbering makes little sense, and the logic of item numbering should be changed...

Few wishes:

- Add an option "Enable Item Numbering In All Folders" in the program's settings. It is mandatory that:
     - Number=yes is specified in the properties of all folders.
     - The number is immediately displayed for existing items in the folders.
     - The item number does not depend on the creation time. The topmost element in any folder will have number 1.
    - If Number=no defined in properties of the folder, then it takes precedence over "Enable Item Numbering For All Folders".
- Add a new parameter for folders called reverse_numbering=1. To simply reverse the numbering order, taking into account new numbering logic. The bottommost item in the folder will have number 1.
- Make the "Select Item by Typing" option in the program's settings optional.
- ~~Add a sub-option called "Select Item in folder by Typing Its Number"~~. 
- There is already a sub-option "Perform the paste operation when the finalized item is a data".

I honestly have no idea what to do if searching by numbers is activated in the root folder with item numbering enabled. This is bad. I thought about it for a long time and that's why i crossed out the questionable wish.
 
b) 1-0 hotkey Pasting

> What about the following specifications?

I think something like this:

- Add a new parameter for folders called number_pasting=true/false
- Add a new parameter for folders called number_expandable=true/false
- 1-0 hotkey Pasting activates if parameter "number_pasting=true" is specified in the properties of any folder.
- The keys 1-0 are registered as a hotkey to paste one of the first 10 items in currently opened folder with "number_pasting=true" parameter.
- The keys are not global and work only in the data tree view.
- The 1-0 key will not work for the "Select Item by Typing" function in the data tree view.
- If the folder has "number_expandable=true" parameter and the item pointed to by the 1-0 key is not data but a folder, the folder expands.
- If the folder has "number_expandable=false" parameter and the item pointed to by the 1-0 key is not data but a folder, nothing happens.

**Reply by itagagaki** [reply]
> there's a bug with folder naming inside History Folder. For some reason they all called "????"

Couldn't reproduce on my end. Please let me know the steps to reproduce.

**Reply by itagagaki** [reply]
Thank you for your thoughts on the details. They are very helpful.

However, I too am still confused on some points. And, there are some inconsistencies that perhaps the original author may not have been aware of.

So let's clear them up one by one.

First, regarding the numbering of items that are automatically added to the history folder with Number=yes.
This is the serial number that is added to the item's display label.
This serial number is 0 when Charu starts and should be incremented by 1 each time text copied to the clipboard is added to the history folder(s). The increment is 1 even if there are multiple history folders.

There is indeed a code that resets this serial number to 0 when the Count in the history folder is exceeded. But there seems to be a mistake in the timing. And if there is more than one history folder, this idea is broken.

I don't think it's necessary to reset this serial number at all. However, it would be nice to have a "reset history number" button in the menu.

And I do not think this history number has nothing to do with the hotkey feature. What do you think about this?

(Also, the history number is formatted as "`%02d`", but I think "`%d`" is fine. I noticed this in your screenshot).

**Reply by Lichtenshtein** [reply]
> Couldn't reproduce on my end. Please let me know the steps to reproduce.

Steps to reproduce:

- Download the release version.
- Create a history folder inside the data tree.
- Open the properties of the history folder and enter:
     - Count=20
     - ClassHistory=5
     - DuplicationCheck=no
- Copy the text 6 times.
- Font settings in Charu settings do not matter.

I am experiencing this in the clean release version 3.8.3 and in 3 test versions:

Charu3_20230921
Charu3_20230914
Charu3_20230908

Perhaps i didn't notice it immediately because i transferred my data tree, where subfolders in the history folder were already created with the correct names.

In the original Charu, subfolders in the history folder are still created with the correct names.

**Reply by Lichtenshtein** [reply]
> history number is formatted as "%02d", but I think "%d" is fine

Do you mean "1" instead of "01"? Isn't it done for alignment? If the lines of text are out of place at "9-10", it will no longer look aesthetically pleasing. If the text will not stand out, then let it be according to your opinion.

> I do not think this history number has nothing to do with the hotkey feature. What do you think about this?

History number = Serial number +1? 

Do you want to bind any of the potential functions to the Serial number? Which one?

"Select and Paste Item in folder by Typing Its Number", "Hot buffer" with global key modifiers, or "1-0 pasting" with non global key modifiers?

> This is the serial number that is added to the item's display label.

If the Serial number is not reseted AND is calculated from the end before being added to the item's display label (then the first item in the folders will have the History Number 1, if i understood correctly), then this may be useful.

Then if implement reverse_numbering=1 property to folders it will make Charu act the old way (number 1 is last in folder).

Oh, History number starts from 0 too. The folder itself does not count. With ClassHistory=22 i have 23 items outside subfolders. Then even if calculate Serial number from the end the to display History Number, the item History number on top of folder will be 0, not 1.

Sorry, i don't think i understand what you're talking about at all. You should know better about this matter.

**Reply by itagagaki** [reply]
Regarding to the folder naming bug in the history folders:

Oh, I see. This is a lack of i18n work. The name of the folder is "階層履歴" in Japanese, but if the Windows display language is not Japanese, it changes to "????". I will fill in the English resource for this part.

However, the literal translation of "階層履歴" is "Hierarchical History", but I do not understand the sense of this naming. I also don't understand the meaning of "Class" in the property name. Should this mechanism be called "Class" or "Hierarchical"?
I feel it should be called something more precise...

**Reply by Lichtenshtein** [reply]
> However, the literal translation of "階層履歴" is "Hierarchical History", but I do not understand the sense of this naming. I also don't understand the meaning of "Class" in the property name. Should this mechanism be called "Class" or "Hierarchical"?

I think this means the same as a "group". A bundle or a bunch of branches in a tree. For count=140, there are a total of 140 branches. With ClassHistory=22, there are 22 branches in the bunch.

I think we should name it based on the verb. What are we doing? We are grouping history. Into what? Into groups of folders. Perhaps there, where i have "????" instead of the folder name, "History Group" would be more suitable than "Class History". The parameter name could be changed from "ClassHistory" to "GroupHistoryBy" to be more understandable.

I can't come up with anything better yet.

**Reply by itagagaki** [reply]
I think your suggestion is very good. Thanks to your thinking I came up with another idea. How about the folder name is "Archived" and the history folder property name is "ArchiveHistoryBy"?

**Reply by Lichtenshtein** [reply]
> How about the folder name is "Archived" and the history folder property name is "ArchiveHistoryBy"?

It's also good, i think.

---
I have redrawn the old-new set of icons for Charu. It is suitable for dark themes. You can use it.

[itemIconNew.zip](https://github.com/itagagaki/charu3-SE/files/13031720/itemIconNew.zip)

---
Some time ago, i posted an announcement about the release of your fork on a russian forum, in the [thread](https://forum.ru-board.com/topic.cgi?forum=5&topic=22&start=900#21) about clipboard managers. I will also occasionally update information about new versions there. I hope you don't mind.

**Reply by itagagaki** [reply]
In Charu3.9.0, the folder name is "Archived history" and the optional property is "ArchiveUnit=N".
The previous property name is also retained for compatibility.

Thanks for the icon. I will add it to the repository later and include it in the next release.

And thank you for promoting Charu. I'm glad! :-)

**Comment by itagagaki** [comment]
@Lichtenshtein 

> 5. An option to disable the search by first letters in Charu window.

Is it to allow the 1-0 key to specify the first 10 objects?
If that's the only reason, wouldn't it be better to add an option that the 1-0 key becomes a function to specify the first 10 objects?

**Reply by Lichtenshtein** [reply]
> Is it to allow the 1-0 key to specify the first 10 objects?

Yes. I thought these functions would conflict if they coexisted together. If one option is selected, the other would automatically turn off.

> If that's the only reason, wouldn't it be better to add an option that the 1-0 key becomes a function to specify the first 10 objects?

I don't understand.

**Comment by itagagaki** [comment]
> 1. ActivatIng/inserting an object with a single mouse click. Does not apply to folders. It might be worth adding a special mode for editing objects in the Charu window to avoid problems with accidental paste triggering while moving objects.

Now possible on master branch.

**Reply by Lichtenshtein** [reply]
> Now possible on master branch.

Holy cow. I didn't notice this change.

Conflicts with "Select Multiple Items" option in Charu Popup... When you check the checkbox, the insertion of the fragment is triggered. Apparently, it is necessary to shift a bit the triggering area in selection mode...

**Reply by itagagaki** [reply]
Thank you for your report. Fixed so that clicking on a checkbox does not trigger it. It will be ok in the next build.

**Comment by itagagaki** [comment]
@Lichtenshtein 

> 2. Keep the focus on the last window when you call Charu. It is increasingly common to see input forms that hide when the focus shifts from them (e.g., browser extensions like Selection Search that work in popups). While selecting an object to insert from Charu the focus is lost and you have to reopen these forms all the time.

I haven't been able to reproduce this issue, so please send me details of your setup. It would be more helpful if you could provide a video.

**Reply by Lichtenshtein** [reply]
> I haven't been able to reproduce this issue, so please send me details of your setup. It would be more helpful if you could provide a video.

I haven't dealt with such types of input forms for a long time. Maybe because out of habit i no longer use them. Because I know that Charu doesn't work with them. I forgot where I encountered them.

I can say for sure that this applies to browser extensions for Chrome. If the extension has a Popup, and the popup has a search form, then Charu will not work with it.

Extension Selection Search for example:

https://chrome.google.com/webstore/detail/selection-search/gipnlpdeieaidmmeaichnddnmjmcakoe?hl=en

You can try inserting something using Charu into the search form shown in the picture:

https://lh3.googleusercontent.com/XsHtOKBYDnlmREBvDEbRaiZdy0XGvIBgJNLh7yELgb1Vsop0eVfEjYm4FsceM9ZTQLAaIouJHdkCSvBWL2Rps4fB=w640-h400-e365-rj-sc0x00ffffff

If something doesn't work for you, i will try to record a gif or video for you.

**Reply by itagagaki** [reply]
Since Charu never changes the focus of other windows, we do not know the cause.
I experimented a bit. Take a look at this recorded video.
The first time, I just enter data from the Charu popup. Then Selection Search does nothing and closes, right?
The second time, after the Charu popup, I manually focus the Selection Search input field again manually (then the Charu popup closes, so), let it pop up again, and enter the data. In this case, the results seem to be as expected.
The third time, I manually focus the Selection Search input field carefully, then let the Charu pop up again, and enter the data. In this case, however, Selection Search does nothing and closes.
All three times Charu just does exactly the same thing, so at this point I have no idea why this add-on is behaves so differently.

https://github.com/itagagaki/charu3-SE/assets/71218184/8aa20da2-f028-4b9f-8311-8323332bf9b6

**Reply by Lichtenshtein** [reply]
Yeah, i managed to do it too. It feels like some strange magic.

This trick works in any chrome extension.

Charu supports these parameters, which can be set in the properties for items (or folders, not checked).

> Macro=   For example, if you specify "Macro=QTClip" in a certain folder, the Charu3 format macros will be disabled in that folder, and only the QTClip format macros will be replaced.
> 
> BeforKey=    You can press a key before pasting the data. The format is the same as "key input macro". If you set it in a folder, it will be effective for all data in that folder.
> Example: BeforeKey=tab;5

It may be possible to come up with something so that the trick always works. I will look into it later.

**Reply by itagagaki** [reply]
Although I have investigated further, this is most likely a browser issue.

Chrome and Edge will close the extension popup when the outside of the popup is clicked or the ESC key is pressed.
However, when I monitored Windows events, no such event occurred. It is also unlikely that the extension closes the popup spontaneously. Most likely a Chrome/Edge bug.

Firefox seems to close the extension popup when it loses focus; it also closes it when clicked outside of Firefox. So even just popping up the Charu it will close it, which is useless...

I will be sending feedback to Google later, and I will also be submitting a feature request to Mozilla via Bugzilla.

**Reply by Lichtenshtein** [reply]
In the very first message i mentioned another clipboard manager, [CopyQ](https://github.com/hluk/CopyQ), in which as a testing i was able to work with such input forms like in Chrome extension popups. However, in the latest version of CopyQ that i just downloaded, this no longer works. The extension popup does not close, and CopyQ somehow can still keep the focus on the window, but not on the text field. It is now impossible to insert a fragment into the text field on the first try. I think this is definitely a browser issue. Or a feature that is designed according to some new security standards or whatever.

About 10 years ago, Chrome users complained that with the introduction of manifest_version 2 for extensions, they had problems with focusing on input fields.

https://stackoverflow.com/questions/11347149/cant-set-focus-to-input-in-chrome-extension

https://bugs.chromium.org/p/chromium/issues/detail?id=111660

Perhaps with the transition to manifest_version 3, Google went further and broke something else. Frankly speaking, this is a common practice for them - discreetly removing or breaking something useful that is used by few users according to their internal statistics. Most users won't pay attention to it, and for the minority, they rarely bring anything back (like tab muting).

I have a biased attitude towards Google, based on their previous actions. Of course, if it's a bug, then it should be fixed by them.

**Reply by itagagaki** [reply]
I have submitted the following feedback to Google.

> There is a problem with focus in Chrome extension popups.
> 
> Steps to reproduce:
> 1. Add an extension to Chrome that has a text input form in the pop-up.
> 2. Click on the extension's icon to display its pop-up and focus on the text input field.
> 3. Launch another Windows GUI application (any. Explorer, for example) or activate an application that is already running, if any. The focus will naturally move to that application.
> 4. Use Alt+Tab to bring the focus back to Chrome.
> 
> Then the text input field will not have the focus. If you press Ctrl+V to paste text from the clipboard here, the extension will close.
> 
> If you do the following between steps 3 and 4 above, the focus will return to the text input field and the paste will take place there.
> 
> Click on the text entry field in the extension popup to refocus it once, and then activate another application window again.

**Reply by itagagaki** [reply]
And I have created an issue in Bugzilla (Mozilla).
https://bugzilla.mozilla.org/show_bug.cgi?id=1855270

**Reply by Lichtenshtein** [reply]
> And I have created an issue in Bugzilla (Mozilla).
[bugzilla.mozilla.org/show_bug.cgi?id=1855270](https://bugzilla.mozilla.org/show_bug.cgi?id=1855270)

I registered, and did not find the "upvote" button. Thought they were everywhere now.

**Reply by itagagaki** [reply]
I see it here.
![image](https://github.com/itagagaki/charu3-SE/assets/71218184/e181992d-5512-41ea-96a9-963571911673)

**Reply by Lichtenshtein** [reply]
Upvoted. Thanks.

**Comment by itagagaki** [comment]
@Lichtenshtein 

> 4. Burst paste.

a) might be possible.  That is, in a single paste operation, the items that have been stocked up to that point are concatenated into a string, and it is copied to the clipboard and pasted, correct?

b) is a little unclear. Does this mean do something similar thing for checkboxed items? If so, in what order should they be concatenated?

**Reply by Lichtenshtein** [reply]
> a) might be possible. That is, in a single paste operation, the items that have been stocked up to that point are concatenated into a string, and it is copied to the clipboard and pasted, correct?

The "Burst paste" function might has been useful to me in many cases. Sometimes, yes, to combine elements into one line and separate them like this:

`test1; test2; test3; test4`

Sometimes like this:

`test1, test2, test3, test4`

More often like this:

```
test1
test2
test3
test4
```

In this case, it is sometimes necessary to separate fragments with an arbitrary separator, like this:

```
test1
; example ;
test2
; example ;
test3
; example ;
test4
```

The last one can be excessive. But the one-time insertion of each word on a *new* line was often necessary.

**ADDITION**. By and large, the difference in all these options 1,2,3,4 - is only in the type of separator. It would be great if it was possible to specify the type of separator (with the ability to wrap the line) somewhere in the options.

> b) is a little unclear. Does this mean do something similar thing for checkboxed items? 

Yes. It means that when you select fragment using the "Select Multiple Items" option in Charu Popup, the new "Burst paste these fragments" option appears active in the context menu.

> If so, in what order should they be concatenated?

Hm. I don't know.

Maybe add a new option "Select Multiple Items for burst paste" to the context menu? Then, when selecting each fragment, they are copied as if in FIFO/FILO mode. The insertion order depends on the selected FIFO or FILO mode in the program options. After selection, we press "Burst paste" in the context menu and they are inserted as in point **a)**.

**Comment by Lichtenshtein** [comment]
> Several independent clipboards with different histories - a request to compare to the new functionality. I don't really need something like that. Although, I have seen similar questions in the topics of other clipboard managers. Such a feature in Charu might attract some new users.

By the way, i learned that Charu already has something like a second clipboard. An _independent_ history folder. In fact, there can be as many as you want. Because any folder can be turned into a new _independent_ History Folder. 

To do this, in the options of any folder, you need to specify "DirectCopyKey=ctrl+f9 (any hotkey)". Then, by pressing this key (ctrl+f9), fragments of the selected text will be copied to this folder just like when copying to a regular History Folder. If you assign a hotkey to this folder, for example Hotkey=shift+f9, then in fact it will become the second clipboard.

Description from the doc:

> DirectCopyKey=When you write ctrl+f9, when you select text in an editor or other application and press ctrl+f9, the selected text will be entered into that data.  
>
> When used in combination with Hotkey macros, it can be used with a sense of having a dedicated clipboard for specific key operations. For example, using DirectCopyKey=ctrl+f9 for copying and Hotkey=shift+f9 for pasting.  
> When used in a folder, it can be used in a way similar to a history folder where you c accumulate history.

**Comment by itagagaki** [comment]
> 7. Automaticaly "unlock" objects copied from specific applications, such as password managers (KeePass, 1Password, etc). Idea is in one-time paste, to not keep passwords in history. Or option to disable adding objects to history for specific programs at all.

I thought about it, there is no way to know which application added the text to the clipboard.

**Reply by Lichtenshtein** [reply]
In CLCL (equivalent of Charu, also japanese) you can specify the name and class of the window of a third-party program in the settings.  

![123](https://github.com/itagagaki/charu3-SE/assets/13855295/fc6e2f4e-3df8-42de-819e-0c6649506406)

The CLCL is open source. I looked for the function that is responsible for this, just to get an idea of how it works. But i'm not a programmer.

It seems to me that this is the enum_windows_proc function here: 
https://github.com/AndreyKaiu/CLCL-EN-RU-JP-Plugins/blob/main/CLCLSet/CLCLSet.c#L318
Here the file seems to be entirely related to this function:
https://github.com/AndreyKaiu/CLCL-EN-RU-JP-Plugins/blob/main/CLCLSet/SetWindow.c

Japanese version of CLCL: https://kaiu.narod.ru/CLCL/clcl_all_plugins_Ver_2_1_3_3_JP.zip

**Reply by itagagaki** [reply]
I took a quick look at the CLCL code and didn't really understand it..

So I thought about it again.
Although Charu can (and already does) get notifications when the clipboard content changes, there is no way to know which program changed it. However, it might be possible to assume that the owner of the currently active window is the actor.

But if someone else has no window and changes the clipboard in the background, there is no way to know..

So my proposal: if the clipboard content changes while a registered window is active, this content will not be recorded in the Charu's history -- does that sound convenient?

**Reply by Lichtenshtein** [reply]
Applications may have child windows that can be interpreted by Charu as not related to the actor window. Depends on the implementation and on the specific application.

> if the clipboard content changes while a registered window is active, this content will not be recorded in the Charu's history -- does that sound convenient?

Yes, it sounds very good :)

**Reply by itagagaki** [reply]
I have implemented this feature. I couldn't think of an appropriate name for it, so I decided to call it the "Stealth Programs list". The list is now in the "Advanced" tab of the Settings.

[Charu3_20231024.zip](https://github.com/itagagaki/charu3-SE/files/13113782/Charu3_20231024.zip)

**Reply by Lichtenshtein** [reply]
> I have implemented this feature. I couldn't think of an appropriate name for it, so I decided to call it the "Stealth Programs list". The list is now in the "Advanced" tab of the Settings.

Great, i will try it :)

What can be entered in the field before "Add"? Just the name of the exe file?

It looks like there is some problem with the duplication check or something

Reproduce steps:

- The DuplicationCheck option in the history folder should be enabled (DuplicationCheck=1)
- Add the program's exe file to the strealth list
- Open this program, copy the text fragment to the clipboard in the active window of this program
- Delete the program's exe file from the strealth list
- Copy the same text fragment to the clipboard again in the active window of this program
- In the History Folder nothing will appear

**Reply by itagagaki** [reply]
Ah, it's a simple bug. I will fix it.
You are very helpful in finding various bugs :-)

You can either add the full pathname of the executable or just the exe filename to the stealth list. Just a filename will match any folder.

**Reply by Lichtenshtein** [reply]
I wish there were more of us than just me. Then we could find more bugs for you (ha-ha).

Should the option in Settings → General → "Use the tooltip on the notification area icon to display clipboard text" display text from programs listed in the Stealth list? On the one hand, it can be convenient. On the other hand, from the point of security of, for example, passwords, it should not. I think Charu should not display content of clipboard for stealthed programs.

**Reply by itagagaki** [reply]
This is the daily version with some modifications.
[Charu3_20231025.zip](https://github.com/itagagaki/charu3-SE/files/13166221/Charu3_20231025.zip)

**Reply by Lichtenshtein** [reply]
😎😎😎😎 Stealthed 😎😎😎😎

Cool.

In my opinion, the function works fine for now, congratulations. :)
